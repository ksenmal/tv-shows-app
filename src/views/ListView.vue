<script setup>
import { onMounted, ref, watch } from "vue";
import { useRoute } from "vue-router";
import { store } from "../utils/store.js";
import fn from "../utils/functions.js";
import ShowsList from "../components/ShowsList.vue";
import GenreList from "../components/GenreList.vue";

const filteredList = ref([]);
const genre = ref("");
const route = useRoute();

function getFilteredShows(genreName) {
  const arr = store.list.filter((item) => {
    if (item.genres && item.genres.length) {
      return item.genres.indexOf(genreName) === -1 ? false : true;
    } else {
      return false;
    }
  });
  return arr;
}

function getPopularShows() {
  return store.list.slice(0, 20);
}

function populateList(genreName) {
  if (genreName) {
    genre.value = fn.getCapitalizedString(genreName);
    filteredList.value = getFilteredShows(genre.value);
  } else {
    genre.value = "";
    filteredList.value = getPopularShows();
  }
}

watch(
  () => route.params.genre,
  (newGenre) => {
    populateList(newGenre);
  }
);

onMounted(() => {
  populateList(route.params.genre);
});
</script>

<template>
  <GenreList />
  <main>
    <h2 v-if="genre">{{ genre }} shows</h2>
    <h2 v-else>Popular shows of all times</h2>
    <ShowsList :list="filteredList" />
  </main>
</template>

<style scoped>
main {
  padding-top: 20px;
}
h2 {
  text-align: center;
  margin-bottom: 1rem;
}
</style>
