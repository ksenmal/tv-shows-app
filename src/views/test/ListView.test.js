import { shallowMount } from "@vue/test-utils";
import { describe, expect, test, vi } from "vitest";
import ListView from "../ListView.vue";

vi.mock("vue-router", () => {
  return {
    useRoute: () => {
      return {
        params: {
          genre: "drama",
        },
      };
    },
  };
});

vi.mock("../../utils/store.js", () => {
  return {
    store: {
      list: [
        { genres: ["drama", "comedy"] },
        { genres: ["comedy"] },
        { genres: [] },
      ],
    },
  };
});

describe("ListView suite", () => {
  test("ListView snapshot", () => {
    const wrapper = shallowMount(ListView);
    expect(wrapper.html()).toMatchSnapshot();
  });
});
