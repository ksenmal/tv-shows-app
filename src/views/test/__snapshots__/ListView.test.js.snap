// Vitest Snapshot v1

exports[`ListView suite > ListView snapshot 1`] = `
"<genre-list-stub data-v-2b1d45bd=\\"\\"></genre-list-stub>
<main data-v-2b1d45bd=\\"\\">
  <h2 data-v-2b1d45bd=\\"\\">Popular shows of all times</h2>
  <shows-list-stub list=\\"\\" data-v-2b1d45bd=\\"\\"></shows-list-stub>
</main>"
`;
